File: SortVision/public/code/bubble/golang/bubbleSort.go
package main

import (
	"fmt"
	"reflect"
	"testing"
)

// bubbleSort sorts an array of integers using the Bubble Sort algorithm.
// It takes a slice of integers and returns the sorted slice.
// The algorithm works by repeatedly swapping adjacent elements if they are in the wrong order.
// This process is repeated until no more swaps are needed, indicating the list is sorted.
//
// Time Complexity:
//   - Best Case: O(n) when the array is already sorted (due to early termination optimization)
//   - Average Case: O(n²) due to nested loops
//   - Worst Case: O(n²) when the array is reverse sorted
//
// Space Complexity: O(1) - sorts in-place without additional memory allocation
func bubbleSort(arr []int) []int {
	// Handle edge cases: empty array or single element
	n := len(arr)
	if n <= 1 {
		return arr
	}

	// Outer loop: traverse through all elements
	// After each iteration, the largest unsorted element "bubbles up" to its correct position
	for i := 0; i < n-1; i++ {
		swapped := false  // Optimization flag to detect early termination

		// Inner loop: compare adjacent elements in the unsorted portion
		// The last i elements are already in place after previous passes
		for j := 0; j < n-i-1; j++ {
			// Swap adjacent elements if they're in the wrong order
			if arr[j] > arr[j+1] {
				arr[j], arr[j+1] = arr[j+1], arr[j]
				swapped = true
			}
		}

		// If no swaps occurred in a full pass, the array is sorted
		if !swapped {
			break
		}
	}
	return arr
}

func main() {
	// Example usage
	arr := []int{64, 34, 25, 12, 22, 11, 90}
	fmt.Println("Original array:", arr)
	sortedArr := bubbleSort(arr)
	fmt.Println("Sorted array:  ", sortedArr)
}

// TestBubbleSort contains test cases for the bubbleSort function
func TestBubbleSort(t *testing.T) {
	tests := []struct {
		name     string
		input    []int
		expected []int
	}{
		{"Empty Array", []int{}, []int{}},
		{"Single Element", []int{5}, []int{5}},
		{"Sorted Array", []int{1, 2, 3, 4}, []int{1, 2, 3, 4}},
		{"Reverse Sorted", []int{4, 3, 2, 1}, []int{1, 2, 3, 4}},
		{"Unsorted Array", []int{64, 34, 25, 12, 22, 11, 90}, []int{11, 12, 22, 25, 34, 64, 90}},
		{"With Duplicates", []int{3, 1, 2, 3, 1}, []int{1, 1, 2, 3, 3}},
		{"All Equal Elements", []int{7, 7, 7, 7}, []int{7, 7, 7, 7}},
		{"Negative Numbers", []int{-5, -1, -3, 0, 4}, []int{-5, -3, -1, 0, 4}},
	}

	for _, tc := range tests {
		t.Run(tc.name, func(t *testing.T) {
			// Create a copy to preserve original test case
			inputCopy := make([]int, len(tc.input))
			copy(inputCopy, tc.input)
			
			result := bubbleSort(inputCopy)
			if !reflect.DeepEqual(result, tc.expected) {
				t.Errorf("For input %v, expected %v but got %v", tc.input, tc.expected, result)
			}
		})
	}
}

File: SortVision/public/code/bubble/r/bubbleSort.r
# Bubble Sort in R
# ----------------
# Bubble Sort is a simple sorting algorithm that repeatedly steps through the list,
# compares adjacent elements, and swaps them if they are in the wrong order.
# This process is repeated until the list is sorted.

bubble_sort <- function(arr) {
  # Get the length of the array
  n <- length(arr)
  # Outer loop for each pass
  for (i in 1:(n-1)) {
    # Inner loop for comparing adjacent elements
    for (j in 1:(n-i)) {
      # If the current element is greater than the next, swap them
      if (arr[j] > arr[j+1]) {
        temp <- arr[j]        # Store current element in temp
        arr[j] <- arr[j+1]    # Move next element to current position
        arr[j+1] <- temp      # Move temp to next position
      }
    }
  }
  # Return the sorted array
  return(arr)
}

# ----------------------
# Test Cases for Bubble Sort
# ----------------------

# Test 1: Unsorted array
arr1 <- c(5, 2, 9, 1, 5, 6)
cat('Test 1 - Input:', arr1, '\n')
cat('Test 1 - Sorted:', bubble_sort(arr1), '\n\n')

# Test 2: Already sorted array
arr2 <- c(1, 2, 3, 4, 5)
cat('Test 2 - Input:', arr2, '\n')
cat('Test 2 - Sorted:', bubble_sort(arr2), '\n\n')

# Test 3: Array with negative numbers
arr3 <- c(-3, 0, 2, -8, 1)
cat('Test 3 - Input:', arr3, '\n')
cat('Test 3 - Sorted:', bubble_sort(arr3), '\n\n')

# Test 4: Array with all equal elements
arr4 <- c(7, 7, 7, 7)
cat('Test 4 - Input:', arr4, '\n')
cat('Test 4 - Sorted:', bubble_sort(arr4), '\n\n')
